scheduler/scheduler.go

```diff
@@ -227,30 +233,117 @@ func (driver *MesosSchedulerDriver) masterDetected(master *mesos.MasterInfo) {
 		}
 
 		driver.MasterPid = pid // save for downstream ops.
+		driver.tryAuthentication()
+	} else {
+		log.Infoln("No master detected.")
+	}
+}
+
+func (driver *MesosSchedulerDriver) tryAuthentication() {
+	if driver.authenticated {
+		// programming error
+		panic("already authenticated")
+	}
+
+	masterPid := driver.MasterPid // save for referencing later in goroutine
+	if masterPid == nil {
+		log.Info("skipping authentication attempt because we lost the master")
+		return
+	}
+
+	if driver.authenticating != nil {
+		// authentication is in progress, try to cancel it
+		driver.closeAuthenticating()
+
+		// force a retry in authenticate()
+		driver.reauthenticate = true
+		return
+	}
 
-		if driver.credential != nil {
-			if err := driver.authenticate(); err != nil {
+	if driver.credential != nil {
+		// authentication can block and we don't want to hold up the messenger loop
+		authenticating := make(chan struct{})
+		go func() {
+			result := &mesos.InternalAuthenticationResult{
+				//TODO(jdef): is this really needed?
+				Success:   proto.Bool(false),
+				Completed: proto.Bool(false),
+			}
+			if err := driver.authenticate(masterPid, authenticating); err != nil {
 				log.Errorf("Scheduler failed to authenticate: %v\n", err)
-				return
+				if err == auth.AuthenticationFailed {
+					result.Completed = proto.Bool(true)
+				}
+			} else {
+				result.Completed = proto.Bool(true)
+				result.Success = proto.Bool(true)
 			}
-
 			func() {
-				driver.lock.Lock()
-				defer driver.lock.Unlock()
-				driver.authenticated = true
+				// protect against panics from already closed channel
+				defer func() { recover() }()
+				close(authenticating)
 			}()
-
-		} else {
-			log.Infoln("No credentials were provided. " +
-				"Attempting to register scheduler without authentication.")
-			// register framework
-			go driver.doReliableRegistration(float64(registrationBackoffFactor))
-		}
+			driver.messenger.Route(context.TODO(), driver.messenger.UPID(), result)
+		}()
+		driver.authenticating = authenticating
 	} else {
-		log.Infoln("No new master detected.")
+		log.Infoln("No credentials were provided. " +
+			"Attempting to register scheduler without authentication.")
+		driver.authenticated = true
+		go driver.doReliableRegistration(float64(registrationBackoffFactor))
 	}
 }
 
+func (driver *MesosSchedulerDriver) closeAuthenticating() {
+	// someone already may have closed it, so recover
+	// (this is pretty hackish)
+	defer func() { recover() }()
```

Comment:
I don't think that would quite solve it. Consider this:
1. master change detected
2. authentication process (1) initiated (async)
3. master change detected again
4. authentication process (2) initiated (async) -- calls `closeAuthenticating()` before resetting `authentication` chan
5. authentication process (1) completes - handleAuthenticationResult() calls `closeAuthenticating()` --- PANIC

I've saw something like this happen in the integration tests.

Furthermore, writing this out has made me realize that we probably need to track the masterId with the authentication process, otherwise an authentication result could be processed successfully for the wrong master.

I can spend some time on this. Perhaps changing the type of `authenticating` to something like this (which is only roughly go-ish):
```go
type authenticationAttempt struct {
  masterPid string
  done chan struct{}
  doneOnce sync.Once
}
func (a *authenticationAttempt) cancel() {
  if a != nil doneOnce.Do(func() { close(done) }) 
}
func (a *authenticationAttempt) isInProgress() bool {
  if a != nil select {
  case <- done: return false
  default: return true
  }
  return false
}
```